File: JAVA/deleteAndEarn.java
import java.util.*;

class Solution7 {
    public int deleteAndEarn(int[] nums) {
        Arrays.sort(nums);

        int maxElement = nums[nums.length - 1];

        int[] freq = new int[maxElement + 1];

        for (int i = 0; i < nums.length; i++) {
            freq[nums[i]]++;
        }

        return totalPoints(freq, 0, new HashMap<Integer, Integer>());
    }

    private int totalPoints(int[] freq, int currentIndex, HashMap<Integer, Integer> memo) {
        if (currentIndex >= freq.length)
            return 0;

        int currentKey = currentIndex;

        if (memo.containsKey(currentKey))
            return memo.get(currentKey);

        int delete = currentIndex * freq[currentIndex] + totalPoints(freq, currentIndex + 2, memo);

        int noDelete = totalPoints(freq, currentIndex + 1, memo);

        memo.put(currentKey, Math.max(delete, noDelete));

        return Math.max(delete, noDelete);
    }
}

File: C++/deleteAndEarn.cpp
#include<bits/stdc++.h>
using namespace std;

class Solution {
public:
    int deleteAndEarn(vector<int>& nums) {
        int n = nums.size();
        
        int maxElement = *max_element(nums.begin(), nums.end());
        //int maxElement = nums[nums.size()-1];
        vector<int> freq(maxElement+1);
        for(auto &it : nums)
            freq[it]++;
        map<int,int> memo;
        return maxPoints(0,freq,memo);
    }
private:
    int maxPoints(int currentIndex,vector<int> &freq,map<int,int> &memo)
    {
        if(currentIndex >= freq.size())
            return 0;
        
        int currentKey = currentIndex;
        
        if(memo.find(currentKey)!=memo.end())
            return memo[currentKey];
        
        int Delete = currentIndex*freq[currentIndex] + maxPoints(currentIndex+2,freq,memo);
        
        int noDelete = maxPoints(currentIndex+1,freq,memo);
        
        memo[currentKey] =  max(Delete,noDelete);
        return max(Delete,noDelete);
        
    }
};